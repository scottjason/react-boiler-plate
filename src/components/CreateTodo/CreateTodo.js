import React from 'react';
import { Input, Form, Container } from './CreateTodo.style';

const CreateTodo = props => {
  /**
  * Input must be declared here so the ref can refer to it
  */
  let todoInput = React.createRef();
  const currentInputVal = () => todoInput.current.value;
  const onSubmit = e => {
    e.preventDefault();
    props.onCreateTodo(currentInputVal())
  };
  
  return(
    <Container>
      <Form onSubmit={onSubmit}>
        <Input
          ref={todoInput}
          type={'text'}
          placeholder='create todo...'
        />
      </Form>
    </Container>
  )
}

export default CreateTodo;