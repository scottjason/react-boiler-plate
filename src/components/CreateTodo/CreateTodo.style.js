import styled from 'styled-components';

export const Container = styled.div`
  position: relative;
  margin: 0 auto;
  width: auto;
  height: auto;
  max-width: 600px;
`;

export const Form = styled.form`
  position: relative;
  margin: 0 auto;
  width: auto;
  height: auto;
  max-width: 600px;
`;

export const Input = styled.input`
  position: relative;
  width: 100%;
  box-sizing: border-box;
  height: 45px;
  padding: 20px;
  font-size: 1.2rem;
  color: #15f4ee;
  background-color: rgba(0, 0, 0, .7);
  border: 1px solid rgba(225, 225, 225, .6);
  outline: 0;
`;
