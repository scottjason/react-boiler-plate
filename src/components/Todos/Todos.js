import React, { Component } from 'react';
import { Container } from './Todos.style';
import Todo from '../Todo/Todo';

class Todos extends Component {
  render() {
    return(
      <Container>
        {this.props.todos.map(todo => {
          return(
            <Todo
              key={todo.id}
              text={todo.text}
            />
          );
        })}
      </Container>
    )
  }
}

export default Todos;