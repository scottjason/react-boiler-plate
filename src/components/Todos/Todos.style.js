import styled from 'styled-components';

export const Container = styled.div`
  position: relative;
  margin: 0 auto;
  width: 70%;
  max-width: 600px;
  height: auto;
  padding: 20px;
  background-color: red;
  font-size: 1.4rem;
  color: white;
`;