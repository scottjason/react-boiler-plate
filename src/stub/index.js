import faker from 'faker';

/* Initial Todos stub data */

const createRandId = () => faker.random.uuid();
const createRandText = () => faker.lorem.sentence();

const stub = {
  todos: [
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
    {id: createRandId(), text: createRandText()},
  ]
}

export default stub;
