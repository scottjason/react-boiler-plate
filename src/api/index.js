/* initial stub data */
import stub from '../stub';

/* extract this method from container for testing purposes */

export const fetchTodos = () => {
  return stub.todos;
};
