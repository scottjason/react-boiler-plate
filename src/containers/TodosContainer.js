import React, { Component } from 'react';
import CreateTodo from '../components/CreateTodo/CreateTodo';
import Todos from '../components/Todos/Todos';
import { fetchTodos } from '../api';
import faker from 'faker';

/* initial stub data */
import stub from '../stub';

class TodosContainer extends Component {

  _isMounted = false;
  
  constructor(props) {
    super();
    this.state = {
      todos: [],
    };
  }
  componentDidMount() {
    this._isMounted = true;

    fetchTodos(stub).then(data => {
      let todos = data.todos;
      if (this._isMounted) {
        todos = todos.map(todo=> {
          todo.isEditMode = false
          return todo;
        })
        this.setState({ todos })
      }
    });
  }
  componentWillUnmount() {
    this._isMounted = false;
  }
  createRandId() {
    return faker.random.uuid();
  }
  
  sortByMostRecent(x, y) {
    return x.createdAt - y.createdAt;
  }
  
  onCreateTodo = text => {
    let todo = {
      id: this.createRandId(),
      text: text,
      isEditMode: false,
      createdAt: Date.now(),
    };
    let todos = this.state.todos;
    todos.push(todo);
    todos = todos.sort(this.sortByMostRecent);
    this.setState({todos});
  } 
  onDeleteTodo = todo => {
    let todos = this.state.todos;
    todos = todos.filter(td => td.id !== todo.id);
    this.setState({todos});
  }
  onToggleUpdate = todo => {
    let todos = this.state.todos.map(td=> {
      if (td.id === todo.id) {
        td.isEditMode = !td.isEditMode;
      } else {
        td.isEditMode = false;
      }
      return td;
    });
    this.setState({todos}); 
  }
  onSaveUpdatedTodo = todo => {
    let todos = this.state.todos.map(td=> {
      if(td.id === todo.id) {
        td.text = todo.text;
      }
      td.isEditMode = false;
      return td;
    }).sort(this.sortByMostRecent);
    this.setState({todos});
  }
  render() {
    return(
      <React.Fragment>
        <CreateTodo 
          onCreateTodo={todo=> this.onCreateTodo(todo)}
        />
        <Todos
          todos={this.state.todos}
          onToggleUpdate={todo=> this.onToggleUpdate(todo)}
          onDeleteTodo={todo=> this.onDeleteTodo(todo)}
          onSaveUpdatedTodo={todo=> this.onSaveUpdatedTodo(todo)}
        />
      </React.Fragment>
    )
  }
}

export default TodosContainer;