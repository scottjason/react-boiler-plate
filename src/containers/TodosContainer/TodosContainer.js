import React, { Component } from 'react';
import Todos from '../../components/Todos/Todos';
import { fetchTodos } from '../../api';

/* initial stub data */
import stub from '../../stub';

class TodosContainer extends Component {
  constructor(props) {
    super();
    this.state = {
      todos: [],
    };
  }
  componentWillMount() {
    fetchTodos(stub).then(data => {
      let todos = data.todos;
      this.setState({ todos })
    });
  }
  render() {
    return(
      <Todos todos={this.state.todos} />
    )
  }
}

export default TodosContainer;