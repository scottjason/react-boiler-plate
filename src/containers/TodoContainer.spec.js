import React from 'react';
import { shallow, mount } from 'enzyme';
import TodosContainer from './TodosContainer';
import Todos from '../components/Todos/Todos';
import CreateTodo from '../components/CreateTodo/CreateTodo';
import stub from '../stub';
import { fetchTodos } from '../api';

describe('TodosContainer', () => {

  test('it should be correct element', () => {
  /* shallow, no children */
    let wrapper = shallow(<TodosContainer />);
    expect(wrapper).toMatchElement(<TodosContainer />);
  });

 test('_mounted should be true when mounted', () => {
  /* shallow, no children */
    let wrapper = shallow(<TodosContainer />);
    expect(wrapper.instance()._isMounted).toEqual(true)
  });
  
  test('renders correctly with no props', () => {
    /* shallow, no children */
    let wrapper = shallow(<TodosContainer />);
    expect(wrapper.exists()).toBe(true)
    expect(wrapper).toMatchSnapshot();;
  });
  
  test('it should render both one CreateTodo component and one Todos Component ', () => {
    /* mount, includes children */
    let wrapper = mount(<TodosContainer />);
    return fetchTodos(stub).then(() => {
      expect(wrapper.children(Todos).length).toEqual(1);
      expect(wrapper.children(CreateTodo).length).toEqual(1);
    });
  });
  
  test('todos should be an array in state ', () => {
    /* mount, includes children */
    let wrapper = mount(<TodosContainer />);
    return fetchTodos(stub).then(() => {
      let state = wrapper.instance().state;
      expect(state.todos).toBeInstanceOf(Array)
    });
  });
});
