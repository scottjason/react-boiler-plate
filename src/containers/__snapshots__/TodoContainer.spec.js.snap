// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodosContainer renders correctly with no props 1`] = `
<Fragment>
  <CreateTodo
    onCreateTodo={[Function]}
  />
  <Todos
    onDeleteTodo={[Function]}
    onSaveUpdatedTodo={[Function]}
    onToggleUpdate={[Function]}
    todos={Array []}
  />
</Fragment>
`;
